### DISCUSSION TITLE
Is it possible to use an automap_base without reflecting relationships?

### DISCUSSION THREAD

**Discussion by nmoreaud** [sqlalchemy/sqlalchemy#10041] [Usage Questions]
Hello
I'm working on a program that handles table per table requests.
To reduce initial reflection time, we have tried to reflect tables lazily, only when they are first requested. The automap metadata is filled incrementally.
However, this code is unstable and often regresses.

I'd like to know if there is a simpler approach to reflect table per table while keeping the ability to use each table like a model class.

Problem in bellow code : if I set `resolve_fks` to `False`, `cnx.base.prepare()` fails with `Foreign key associated with column 'table1.COL1' could not find table 'table2.COL2' with which to generate a foreign key to target column 'COL3`.

```python

def name_for_scalar_relationship(base, local_cls, referred_cls, constraint):
    name = referred_cls.__name__.lower()
    local_table = local_cls.__table__
    if name in local_table.columns:
        newname = name + "_"
        get_logger().warning(f"Already detected name {name} present. Using {newname}")
        return newname
    return name

def generate_relationship(base, direction, return_fn, attrname, local_cls, referred_cls, **kw):
    # we don't need table relationships
    return None

def reflect_db_table(cnx, table_name):
        # table does not exist
        if table_name not in cnx.all_tables:
            return

        # create table metadata
        @sqlalchemy.event.listens_for(cnx.meta_data, "column_reflect")
        def receive_column_reflect(inspector, table, column_info):
            # lowercase column names in models
            column_info['key'] = column_info['name'].lower()

        meta_column_list = []
        try:
            cnx.meta_data.reflect(bind=cnx.engine, views=True, only=[table_name], resolve_fks=True)
        except sqlalchemy.exc.InvalidRequestError:
            get_logger().warning(f"Tried to map table {table_name} with lowercase, trying again with upperscase")
            table_name = table_name.upper()
            cnx.meta_data.reflect(bind=cnx.engine, views=True, only=[table_name], resolve_fks=True)
        if table_name not in cnx.meta_data.tables.keys():
            table_name = f'{cnx.db_info.schema}.{table_name}' if cnx.db_info.schema is not None else table_name
        meta_table = cnx.meta_data.tables[table_name]

        if cnx.db_info.type == DatabaseType.SQLITE.value:
            r = cnx.session.execute(text(f"PRAGMA table_info('{meta_table.name}')"))
            sqlite_col_dict = {c['name']: c for c in r.mappings()}

        for column in meta_table.columns:
            # The automapping process ignores views because it cannot detect their primary key
            # We have to declare manually views primary key (= db_record_id)
            if column.name.lower() == "db_record_id":
                meta_column_list.append(
                    Column(column.name, Integer, primary_key=True, autoincrement="auto", nullable=False, key="db_record_id")
                )
                continue

            if cnx.db_info.type == DatabaseType.SQLITE.value:
                # replace LONGVARBINARY (auto detected as Numeric) with a BLOB type
                sqlite_col = sqlite_col_dict[column.name]
                if sqlite_col['type'] == 'LONGVARBINARY':
                    default = column.default
                    meta_column_list.append(
                        Column(column.name, LargeBinary, nullable=column.nullable, default=default, key=column.name.lower())
                    )
                    continue

            # Other column: created by sqlalchemy metadata reflect without modifications

        # patch automaped table
        for meta_column in meta_column_list:
            meta_table.append_column(meta_column, replace_existing=True)

        # recreate dabasase object containing every tables reflected so far
        # class_registry contains every table generated class
        cnx.class_registry.clear()
        cnx.base = automap_base(metadata=cnx.meta_data, class_registry=cnx.class_registry)
        cnx.base.prepare(
            classname_for_table=lower_classname,
            name_for_scalar_relationship=name_for_scalar_relationship,
            generate_relationship=generate_relationship
        )
```

**Comment by CaselIT** [comment]
Hi,

What version on sqlalchemy are you using? Version 2 improved reflection times significantly. How long does it take to run prepare?

**Comment by nmoreaud** [comment]
I have updated to version 2.0.
Prepare() still takes up to 6s on some local databases if I don't use incremental mechanism.

**Reply by CaselIT** [reply]
Understood. What is the db in use?

**Reply by nmoreaud** [reply]
I'm going to check, I don't remember which db was the slowest

**Reply by nmoreaud** [reply]
Okay in general postgres/oracle/sqlite take 2s at most to prepare, but sqlserver takes 4s the first time (local databases).

**Reply by nmoreaud** [reply]
global prepare

oracle took: 1.0
postgres took: 0.9
mssql took: 3.4
sqlite took: 1.1
sqlite mount took: 5.7

incremental

oracle took: 0.2s
postgres took: 0.2s
mssql took: 0.6s (very variable)
sqlite took: 0.1s
sqlite mount took: 0.1s

**Reply by CaselIT** [reply]
ok, if this is a cli it may be annoying, but for a long running application those startup times seem reasonable.

Thanks for checking and providing the timing. Just wanted to ensure that there wasn't something unexpected that was taking minutes or more to reflect

**Comment by zzzeek** [comment]
consider generating models as static code using [sqlacodegen](https://pypi.org/project/sqlacodegen/) ?   automap is useful for its expediency but reflection-at-startup for high volume, on demand production applications is likely not a good idea

**Reply by nmoreaud** [reply]
So I guess there is no easy workaround to avoid relationship detection.
I think that we may come back to the simpler "prepare" version.

Static code is also an option and we use it in others parts of the project, but we still need to uses specific reflection code because our customers often add custom table and fields to the databases.

**Reply by zzzeek** [reply]
certainly if you dont need relationships, use "prepare".  automap is all about trying to get the relationships in.   prepare is simpler, if that does what you need, use that

**Comment by nmoreaud** [comment]
With `resolve_fks=False`, `reflect` still creates a set of foreign keys.
We can clear this set with:
```python
cnx.meta_data.reflect(bind=cnx.engine, views=True, only=[table_name], resolve_fks=False)
meta_table = cnx.meta_data.tables[table_name]
meta_table.foreign_keys.clear()
```

Instead of automap, we can use declarative_base with manually reflected tables: https://docs.sqlalchemy.org/en/14/orm/declarative_tables.html#mapping-declaratively-with-reflected-tables
Ex: 
```python
        # patch automaped table
        for meta_column in meta_column_list:
            meta_table.append_column(meta_column, replace_existing=True)

        if cnx.base is None:
            cnx.base = declarative_base()

        TableNameModel = type(meta_table.name, (cnx.base,), {'__table__': meta_table})
        # <=>
        # class TableNameModel(cnx.base):
        #    __table__ = meta_table

        cnx.class_registry[table_name.lower()] = TableModel
```
